naga/src/back/msl/writer.rs

```diff
@@ -3356,6 +3415,17 @@ impl<W: Write> Writer<W> {
         }
         writeln!(self.out)?;
 
+        writeln!(
+            self.out,
+            r#"
+template<typename R, typename T, typename A> R __make_atomic_cas_result(device A* ptr, T cmp_, T v) {{
+    T cmp = cmp_;
+    bool exchanged = {NAMESPACE}::atomic_compare_exchange_weak_explicit(ptr, &cmp, v, {NAMESPACE}::memory_order_relaxed, {NAMESPACE}::memory_order_relaxed);
+    return R{{cmp, exchanged}};
+}}
+        "#
+        )?;
+
```

Comment:
Please don't just write this definition into every shader whether it's needed or not. We should only generate this code if we actually use it.

You can easily determine whether the module needs this definition by checking whether `PredeclaredType::AtomicCompareExchangeWeakResult` is present in `module.special_types.predeclared_types`, since that will always be there, since the `AtomicResult` expression uses it.

Reply:
Coolio, I didn't know that existed, I'll look into that once I get a chance!